### ISSUE TITLE
Unaligned address passed to AO_load in GC_mark_thread_local_fls_for on m68k

### ISSUE THREAD

**Issue by ivmai** [ivmai/bdwgc#329]
Reported by @hpreusse (https://github.com/vectorgraphics/asymptote/issues/172, https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=968265):

I'm trying to build the package for m68k inside a qemu vroot. The package
fails to build in the test suite, log is attached. Due to this issue we
(probably) get an FTBFS on package asymptote. I tried already the
latest github checkout of libgc, problem still present.

Hilmar
-- System Information:
Debian Release: bullseye/sid
  APT prefers unstable
  APT policy: (500, 'unstable'), (1, 'experimental')
Architecture: i386 (i686)

Kernel: Linux 5.7.0-1-686-pae (SMP w/2 CPU threads)
Locale: LANG=de_DE.UTF-8, LC_CTYPE=de_DE.UTF-8 (charmap=UTF-8) (ignored: LC_ALL set to en_GB.UTF-8), LANGUAGE not set
Shell: /bin/sh linked to /bin/dash
Init: systemd (via /run/systemd/system)

Version: 1:8.0.4-2

/usr/include/atomic_ops/sysdeps/loadstore/atomic_load.h:31: AO_load: Assertion `((size_t)addr & (sizeof(*addr) - 1)) == 0' failed.
qemu: uncaught target signal 6 (Aborted) - core dumped

(gdb) bt
\#0  __GI_raise (sig=6) at ../sysdeps/unix/sysv/linux/raise.c:50
\# 1  0xfd9c85e8 in __GI_abort () at abort.c:79
\# 2  0xfd9d4142 in __assert_fail_base (fmt=0xfdacba4b "%s%s%s:%u: %s%sAssertion `%s' failed.\n%n", assertion=0xfe1dce6d "((size_t)addr & (sizeof(*addr) - 1)) == 0",
    file=0xfe1dce97 "/usr/include/atomic_ops/sysdeps/loadstore/atomic_load.h", line=31, function=0xfe1de77d <__PRETTY_FUNCTION__.0> "AO_load") at assert.c:92
\# 3  0xfd9d41ca in __GI___assert_fail (assertion=0xfe1dce6d "((size_t)addr & (sizeof(*addr) - 1)) == 0", file=0xfe1dce97 "/usr/include/atomic_ops/sysdeps/loadstore/atomic_load.h", line=31,
    function=0xfe1de77d <__PRETTY_FUNCTION__.0> "AO_load") at assert.c:101
\# 4  0xfe1dba0e in AO_load (addr=<optimized out>) at /usr/include/atomic_ops/sysdeps/loadstore/atomic_load.h:31
\# 5  AO_load (addr=<optimized out>) at /usr/include/atomic_ops/sysdeps/loadstore/atomic_load.h:28
\# 6  GC_mark_thread_local_fls_for (p=0xfe2133e2 <first_thread+54>) at thread_local_alloc.c:271
\# 7  0xfe1d854e in GC_mark_thread_local_free_lists () at pthread_support.c:284
\# 8  0xfe1d1cec in GC_push_roots (all=1, cold_gc_frame=0xfeb59a68 "\376\034b\n\376\265\232h") at mark_rts.c:938
\# 9  0xfe1d10aa in GC_mark_some (cold_gc_frame=0xfeb59a68 "\376\034b\n\376\265\232h") at mark.c:355
\# 10 0xfe1c620a in GC_stopped_mark (stop_func=0xfe1c5be4 <GC_never_stop_func>) at alloc.c:793
\# 11 0xfe1c6986 in GC_try_to_collect_inner (stop_func=0xfe1c5be4 <GC_never_stop_func>) at alloc.c:541
\# 12 GC_try_to_collect_inner (stop_func=0xfe1c5be4 <GC_never_stop_func>) at alloc.c:485
\# 13 0xfe1d36e2 in GC_init () at misc.c:1325
\# 14 GC_init () at misc.c:901

**Comment by ivmai** [comment]
The issue is caused that the structure field (GC_Thread_Rep.tlfs, p=0xfe2133e2) is not 4-byte aligned on m68k, see https://github.com/ivmai/libatomic_ops/blob/master/src/atomic_ops/sysdeps/gcc/m68k.h#L21

**Comment by ivmai** [comment]
See also https://trofi.github.io/posts/191-ghc-on-m68k.html

I'm thinking how to fix it

**Comment by ivmai** [comment]
Could you please try if the following temporary workaround really works:

```
diff --git a/include/private/pthread_support.h b/include/private/pthread_support.h
index 8e4bccf7..fe553775 100644
--- a/include/private/pthread_support.h
+++ b/include/private/pthread_support.h
@@ -93,14 +93,15 @@ typedef struct GC_Thread_Rep {
                                 /* not need to be sent a signal to stop */
                                 /* it.                                  */
 
-    unsigned short finalizer_skipped;
     unsigned char finalizer_nested;
+    unsigned short finalizer_skipped;
                                 /* Used by GC_check_finalizer_nested()  */
                                 /* to minimize the level of recursion   */
                                 /* when a client finalizer allocates    */
                                 /* memory (initially both are 0).       */
 
-    ptr_t stack_end;            /* Cold end of the stack (except for    */
+    ptr_t stack_end __attribute__((__aligned__(sizeof(word))));
+                                /* Cold end of the stack (except for    */
                                 /* main thread).                        */
     ptr_t altstack;             /* The start of the alt-stack if there  */
                                 /* is one, NULL otherwise.              */
```

If yes, I will polish it.

**Comment by ivmai** [comment]
Should be fixed by b55459a, please check.

**Comment by glaubitz** [comment]
> Should be fixed by [b55459a](https://github.com/ivmai/bdwgc/commit/b55459ad47d9aaa75c9a4b0f99a3b867592ab8e0), please check.

This fixes the issue for me.
